.github/workflows/yocto-build-deploy.yml

```diff
@@ -0,0 +1,639 @@
+
+    defaults:
+      run:
+        working-directory: .
+        shell: bash --noprofile --norc -eo pipefail -x {0}
+
+    steps:
+      # TODO: clone balena-yocto-scripts @ yocto_scripts_ref instead
+      # https://github.com/actions/checkout
+      - name: Clone device repository
+        uses: actions/checkout@v3
+        with:
+          repository: ${{ inputs.device-repo || github.repository }}
+          token: ${{ secrets.GITHUB_TOKEN }}
+          ref: ${{ inputs.device-repo-ref || github.ref }}
+          submodules: true
+
+      - name: Update balena-yocto-scripts submodule to ${{ inputs.yocto-scripts-ref }}
+        if: inputs.yocto-scripts-ref != ''
+        working-directory: ./balena-yocto-scripts
+        run: |
+          git config --add remote.origin.fetch '+refs/pull/*:refs/remotes/origin/pr/*'
+          git fetch --all
+          git checkout --force "${{ inputs.yocto-scripts-ref }}"
+          git submodule update --init --recursive
+
+      # https://github.com/actions/download-artifact
+      - name: Download build artifacts
+        uses: actions/download-artifact@eaceaf801fd36c7dee90939fad912460b18a1ffe # v4.1.2
+        with:
+          path: ${{ runner.temp }}
+          name: build-artifacts
+
+      - name: Extract build artifacts
+        run: |
+          tar -xvf ${{ runner.temp }}/build-artifacts.tar.zst
+          find .
+
+      - name: Set s3 outputs
+        id: s3-lib
+        run: |
+          echo "s3_policy=private" >> $GITHUB_OUTPUT
+          if [ "${{ needs.build.outputs.is_private }}" = "false" ]; then
+            echo "s3_policy=public-read" >> $GITHUB_OUTPUT
+          fi
+
+          echo "deployer_uid=$(id -u)" >> $GITHUB_OUTPUT
+          echo "deployer_gid=$(id -g)" >> $GITHUB_OUTPUT
+
+          s3_bucket_path="${{ inputs.s3-bucket}}/images"
+          if [ "${{ inputs.deploy-esr}}" = "true" ]; then
+            s3_bucket_path="${{ inputs.s3-bucket}}/esr-images"
+          fi
+
+      # https://github.com/docker/setup-buildx-action
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      # login required to pull private balena/balena-img image
+      # https://github.com/docker/login-action
+      - name: Login to Docker Hub
+        uses: docker/login-action@343f7c4344506bcbf9b4de18042ae17996df046d # v3.0.0
+        with:
+          registry: docker.io
+          username: ${{ secrets.DOCKERHUB_USER }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      # https://github.com/docker/build-push-action
+      - name: Build helper image
+        id: helper-image
+        uses: docker/build-push-action@v5.1.0
+        with:
+          context: "${{ github.workspace }}/balena-yocto-scripts/automation"
+          file: Dockerfile_s3-deploy-env
+          push: false
+          build-args: |
+            DEPLOYER_UID=${{ steps.s3-lib.outputs.deployer_uid }}
+            DEPLOYER_GID=${{ steps.s3-lib.outputs.deployer_gid }}
+
+      - name: Deploy to S3
```

Comment:
Maybe we should look for a s3 deploy action that is maintained externally. Also, I'd probably push for file uploads to be used instead of wasting time and energy on this.

Reply:
AFAIK the balena-s3 image actually handles the preparation of the deflate files and not just the uploading to S3. If that's not the case, then why did we bother using balena-s3 in the first place?

IIRC there is a bunch of special handling in balena-S3 for creating the required artifacts in the prepare.ts script before we start uploading.